documentation/rfc7700.md

```diff
@@ -0,0 +1,151 @@
+
+
+## Casemapping
+
+The server indicates that this casemapping is in use by advertising the `"CASEMAPPING"` `RPL_ISUPPORT` parameter with the value `"rfc7700"`. An example of this is shown below:
+
+    :irc.example.com 005 dan CASEMAPPING=rfc7700 :are supported by this server"
+
+When the `rfc7700` casemapping is in use, names (and other messages from the IRCd where possible) MUST be sent using the [UTF-8](https://tools.ietf.org/html/rfc3629) encoding.
+
+The `rfc7700` casemapping uses the PRECIS [Nickname profile](https://tools.ietf.org/html/rfc7700#section-2) as defined in [Section 2 of RFC 7700](https://tools.ietf.org/html/rfc7700#section-2).
+
+
+### Order of Operations
+
+Names being prepared MUST apply the following rules in the order shown:
+
+1. Preperation using the PRECIS [Nickname profile](https://tools.ietf.org/html/rfc7700#section-2).
+2. Check for restricted characters.
+
+These steps MUST happen in the order shown, or else the restricted characters check may miss characters that should be legitimately restricted.
+
+If a name does contain a restricted character (whether disallowed by the [Nickname profile](https://tools.ietf.org/html/rfc7700#section-2.2) or this document), it MUST be rejected by the server and MUST NOT be propogated to other clients. This is done through the appropriate numeric for the command which tried to set or use the invalid name such as `ERR_ERRONEUSNICKNAME`, `ERR_NOSUCHCHANNEL`, or whichever numeric is most appropriate.
+
+
+### Comparisons
+
+When comparing user names or channel names to check for equivalency, both strings are casemapped with the above process. After this, they are compared with a standard bit-for-bit comparison.
+
+
+## Restricted Characters
+
+Certain characters mean special things to the protocol. This section details characters that are not allowed in specific name types.
+
+
+### All Names
+
+Nicknames, usernames, hostnames and channel names cannot contain the following characters:
+
+* `(' ', 0x20)` - Separates parameters.
+* `(':', 0x3A)` - Separates trailing parameter.
+* `(',', 0x2C)` - Used as a separator.
+* `('*', 0x2A)` - Used in mask matching.
+* `('?', 0x3F)` - Used in mask matching.
+
+To be clear, these characters will also be repeated below.
+
+#### Protocol Framing
+
+These characters are disallowed either by the protocol itself or because of how they are used in protocol framing. These won't be repeated below, as sane implementations either break the message on or disallow the character in streams:
+
+* `('\0', 0x00)` - Disallowed by the protocol.
+* `('\n', 0x0A)` - Used in message framing.
+* `('\r', 0x0D)` - Used in message framing.
+
+
+### Nicknames
+
+Nicknames cannot contain the following characters:
+
+* `(' ', 0x20)` - Separates parameters.
+* `(':', 0x3A)` - Separates trailing parameter.
```

Comment:
Hmm, that's fair. In that case, I can just note that the first letter of one can't be `:`? (since if i.e. a nickname started with `:` then you wouldn't be able to use it in normal messages)